### ISSUE TITLE
Question: is this for geometry or raster data?

### ISSUE THREAD

**Issue by rafaqz** [alex-s-gardner/FastGeoProjections.jl#1]

**Comment by alex-s-gardner** [comment]
Right now I'm focusing on generic point transformations in pure Julia... my goal is simply to replace Proj in places with pure julia. Thoughts?

**Comment by rafaqz** [comment]
Ah ok ambitious! For what scope of projections are you thinking?

And yes you only really need points, this will do the rest on  all GeoInterface.jl compatible objects:
https://github.com/asinghvi17/GeometryOps.jl/blob/main/src/transformations/reproject.jl

Currenly its using Proj.jl to transform points (as it walks the geometries) but you could swap in another function just by removing the dispatch on Proj.Transform.

**Comment by alex-s-gardner** [comment]
@rafaqz my first crack at polar stereo is not all that promising...  easy to match Proj speed ... hard to beat

**Comment by rafaqz** [comment]
You might need to do them in huge batches on GPUs to beat Proj.jl ;)

**Comment by alex-s-gardner** [comment]
With LoopVectorization and code refactoring we can get a 10X+ speedup

**Comment by rafaqz** [comment]
Be sure to check your TTFX with LoopVectorisation.jl, maybe its better with 1.9 but you may need some precompile statements and to reduce type complexity to a minimum

**Comment by asinghvi17** [comment]
For raster data this is very promising.  But I'm not sure how we could apply LoopVectorization enhancements to the point-by-point case.

It would be nice to have these objects be CoordinateTransformations.jl objects, callable on single points, with specializations for broadcast that call the optimized methods.  Not sure how much complexity that introduces though, but at least then we have a way to test that does not involve calling Proj.

Potentially the "single-point-transform" function can be moved outside and take parameters, then the broadcast implementation could just call it?  Something like this:

https://github.com/alex-s-gardner/FastGeoProjections.jl/blob/d19c373c89cc7ef2a7e0b14861e231d32e26cebf/src/polarstereo.jl#L60-L65

can be a function that takes in all the params that are calculated up top, and another function can compute all those parameters.  Then it's completely reusable and we can also try to force inlining.

Parameters that do not require the lat/long coords to calculate can also be precalculated and saved in the coordinate transformation struct, making it significantly easier...

**Comment by alex-s-gardner** [comment]
I agree. FastGeoProjections should operate on single points and should be GeoInterface compatible.

But I wonder what the long-term goal for FastGeoProjectons is. Most of the work in developing a library like this is in the translation of the transformation functions with has mostly been done in [CoordRefSystems.jl](https://github.com/JuliaEarth/CoordRefSystems.jl)

If we really want to replace Proj.jl then we would need to either pull from CoordRefSystems.jl or transcribe a large number of transformations... or just take the hit and stick with Proj.jl

**Comment by alex-s-gardner** [comment]
Both, and the design settled the way this thread suggested it should.

A `Transformation` is a point operator — a callable struct mapping one
point to one point, holding everything the projection precomputes. That
same operator is what `transform`/`transform!` apply to whole
collections, on SIMD lanes and across threads, so there is one
implementation of each projection rather than one per calling
convention. Point-by-point is the primitive; the array path is built on
it, not beside it.

It subtypes `CoordinateTransformations.Transformation` and is
GeoInterface-compatible (#22), so a raster caller and a geometry caller
reach the same code.

Closing as answered.

Co-authored-by: Claude <noreply@anthropic.com>
